### PR TITLE
Adding a phase for synchronizing defTree

### PR DESCRIPTION
**Pull request by EnzeXing**
This PR will add a compiler phase to synchronize defTrees for all definition trees, which is necessary in the initialization checker to access local variables by evaluating their synchronized defTrees

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/core/Phases.scala [hunk 1]
```diff
@@ -290,9 +294,16 @@ object Phases {
...
     def allowsImplicitSearch: Boolean = false
 
-     /** List of names of phases that should precede this phase */
+    /** List of names of phases that should precede this phase */
     def runsAfter: Set[String] = Set.empty
 
+    /** Whether this phase require synchronization of symbol.defTree?
+     *
+     *  The phase `DefTreeSync` will be inserted immediately before this phase
+     *  if `synchronizeDefTree == true`.
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/core/Phases.scala:303]:
```suggestion
     *  if `synchronizeDefTree == true && isEnabled == true`.
```

File: compiler/src/dotty/tools/dotc/core/Phases.scala [hunk 2]
```diff
@@ -105,6 +105,10 @@ object Phases {
               prevPhases += phase.phaseName
               phase
             }
+
+          // synchronize symbol.defTree if necessary
+          if phaseToAdd.synchronizeDefTree then
+            fusedPhases += new SyncDefTree
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/core/Phases.scala:111]: Improvements possible here: if phase `A` and `B` both require `SyncDefTree`, but there are no other phases between `A` and `B`, then we only need to insert one `SyncDefTree` phase before `A`. This avoids inserting a useless `SyncDefTree` before phase `B`.

One way to do that is to introduce another flag `val isCheckerPhase: Boolean`.

File: compiler/src/dotty/tools/dotc/core/Phases.scala [hunk 3]
```diff
@@ -290,9 +294,16 @@ object Phases {
...
 
-     /** List of names of phases that should precede this phase */
+    /** List of names of phases that should precede this phase */
     def runsAfter: Set[String] = Set.empty
 
+    /** Whether this phase require synchronization of symbol.defTree?
+     *
+     *  The phase `DefTreeSync` will be inserted immediately before this phase
+     *  if `synchronizeDefTree == true`.
+     */
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/core/Phases.scala:304]: Add another note:
```
For the synchronization to work reliably, the phase that requires the synchronization
should be in a phase group that does not contain other phases that mutate the trees.

The simplest way to achieve this is to put the phase in a separate phase group.
Compiler plugin phases are each in a separate phase group thus satisfy the requirement.
Having multiple read-only phases inside a phase group is also safe.
```

**Comment by liufengyun** [comment]
@EnzeXing Some init tests have different results after this change, need to reclassify them.

File: compiler/src/dotty/tools/dotc/core/Phases.scala (hunk 2 above)
Comment (olhotak) [reply compiler/src/dotty/tools/dotc/core/Phases.scala:111]: Is the meaning of `isCheckerPhase` that a phase does not modify trees?

If so, should we then go further and enforce this with a new subclass of `Phase` whose `run` method does not update `ctx.compilationUnit.tpdTree`?

**Review by odersky — CHANGES_REQUESTED** [review]
I feel this will not work under incremental compilation. Say, a later phase synchronizes DefTrees. Then during the next run, the frontend phases will see the transformed DefTrees, not the original ones. 

If we do need transformed DefTrees, one could use a Map instead. And this could be done as a preparatory step in the phase that needs them. Which means we do not need to further complicate phase assembly.

So overall, I think we should try a completely different approach.

**Comment by liufengyun** [comment]
Thanks for the feedback @odersky. Doing it inside a phase with a map is a good idea. 

@EnzeXing @olhotak : I'll propose a new fix for the next meeting.

**Pull request closed without merging**